testutil/promrated/prometheus.go

```diff
@@ -0,0 +1,61 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package promrated
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/obolnetwork/charon/app/log"
+)
+
+// ListenAndServe creates a liveness endpoint and serves metrics to prometheus.
+func ListenAndServe(ctx context.Context, addr string) {
+	serverErr := make(chan error, 2)
+	go func() {
+		mux := http.NewServeMux()
+
+		mux.Handle("/livez", http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+			writeResponse(w, http.StatusOK, "ok")
+		}))
+
+		mux.Handle("/metrics",
+			promhttp.Handler(),
+		)
+
+		// Copied from net/http/pprof/pprof.go
+		server := http.Server{Addr: addr, Handler: mux, ReadHeaderTimeout: time.Second}
+		serverErr <- server.ListenAndServe()
+	}()
+
+	for {
+		select {
+		case err := <-serverErr:
+			panic(err)
```

Comment:
panics are to be avoided. Suggest returning the http server from this function and then having your main goroutine in your main function do a single "run everything and wait for shutdown".